Add logoutSilently

majelbstoat opened the pull request:
By submitting a PR to this repository, you agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md). Please see the [contributing guidelines](https://github.com/auth0/.github/blob/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description

This PR adds logoutSilently(), which behaves the same as logout(), but uses runIframe instead of setting window.location to log out.

It also adds cache.delete() to remove a token from the local cache.

I modeled tests on a combination of getTokenSilently and logout.

Feel free to edit this directly if you need to.

### References

Fixes: #308 

### Testing

Call logoutSilently, instead of logout, and confirm that cookies, local storage and cache are cleared as expected.

- [x] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [x] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `master`

majelbstoat commented:
cc: @stevehobbsdev

majelbstoat commented:
The failing test seems to be unrelated?

stevehobbsdev commented:
@majelbstoat Yes, seems to be.

We're about to do a new minor release. Let me get that out of the way, then you can rebase this PR and try again. Might clear up some issues.

majelbstoat commented:
@stevehobbsdev: I rebased, but it still seems to be failing on the same test - something about not running in a secure environment?

stevehobbsdev commented:
Yes, I saw that, looks to be an issue with our tests but I haven’t seen this locally.

Thanks for rebasing, will get a look at this on Monday 👍

stevehobbsdev commented on src/Auth0Client.ts:
````diff
@@ -678,10 +686,59 @@ export default class Auth0Client {
 
+  /**
+   * ```js
+   * auth0.logout();
````
Should be `auth0.logoutSilently` to match the function name.

stevehobbsdev commented on src/Auth0Client.ts:
```diff
@@ -678,10 +686,59 @@ export default class Auth0Client {
+   *
+   * @param options
+   */
+  public async logoutSilently(options: LogoutSilentlyOptions) {
```
Can we move this up the file and group it underneath the default `login` method?

stevehobbsdev replied:
Also, this function should be able to take no arguments and default to an empty object, just like `logout`. Currently I get an error when I pass no arguments.

Can we change that, and also make sure there is a test around that scenario?

stevehobbsdev commented on src/Auth0Client.ts:
```diff
@@ -678,10 +686,59 @@ export default class Auth0Client {
+  public async logoutSilently(options: LogoutSilentlyOptions) {
+    const {
+      federated,
+      localOnly,
```
Let's remove the ability to specify `localOnly` here. It would have the exact same function as in `logout({ localOnly: true })`, but now we're giving developers two ways to accomplish this, which could be confusing.

stevehobbsdev commented on src/global.ts:
```diff
@@ -241,7 +247,25 @@ export interface GetTokenSilentlyOptions {
+   */
+  localOnly?: boolean;
+}
+export interface LogoutOptions extends CommonLogoutOptions {
```
Blank line between the end of the previous interface and this new one, please.

stevehobbsdev commented on src/global.ts:
```diff
@@ -258,25 +282,22 @@ export interface LogoutOptions {
   client_id?: string;
-
+}
+export interface LogoutSilentlyOptions extends CommonLogoutOptions {
```
Blank line between the end of the previous interface and this new one, please.

stevehobbsdev commented on src/utils.ts:
```diff
@@ -347,7 +347,7 @@ export const validateCrypto = () => {
     throw new Error(`
       auth0-spa-js must run on a secure origin.
-      See https://github.com/auth0/auth0-spa-js/blob/master/FAQ.md#why-do-i-get-auth0-spa-js-must-run-on-a-secure-origin 
+      See https://github.com/auth0/auth0-spa-js/blob/master/FAQ.md#why-do-i-get-auth0-spa-js-must-run-on-a-secure-origin
```
The test is failing because of this removal of the space at the end of the line. Can we add it back in?

majelbstoat commented:
rebased and updated the tests to reflect the reduced options.

sorry for all the whitespace and comma changes 😕 do you have an eslint/prettier standard? vscode is kind of militant on altering everything on save.

stevehobbsdev commented:
@majelbstoat No worries. We upgraded to Prettier 2.0 which we've just realised defaults trailing commas to true. We decided to keep that change but it does add a bit of noise to the PRs.

Thanks for the changes - will have a review and let you know.

stevehobbsdev commented:
@majelbstoat I think we're going to need to change up the mechanism slightly. In testing `logoutSilently`, the 60 timeout will always be hit.

The way `runIframe` works is that it relies on Auth0 sending back a response that has some JavaScript in it that uses `postMessage` to return some data back to the SDK. We don't have this for the logout endpoint (it just returns a redirect back to the app), so `runIframe` never knows when it is supposed to complete.

I think we can use the logout endpoint as a kind of fire-and-forget call as there won't be any real useful response. Then we can clear the cache and log the user out locally. Unfortunately it probably means you can't use the `runIframe` method we have (or put a really short timeout on it and absorb the timeout error - bit nasty but would work).

Make sense?

I'd like to help push some changes here but I don't have permission to edit this PR it seems (when I try to push I get asked for my creds).

stevehobbsdev commented:
@majelbstoat I've just merged a change which resets our Prettier config back to some pre-2.0 defaults, if you wouldn't mind re-saving your changed files to match.

majelbstoat commented:
No worries. I'm happy to do it as fire and forget. Before I go do that, this is what I've been using in my own code. Happy with this as an approach?

```
const logoutAuth0Remote = (domain: string) => {
  // Clear the session on auth0.com with a silent iframe.
  const iframe = document.createElement('iframe')
  iframe.setAttribute('width', '0')
  iframe.setAttribute('height', '0')
  iframe.style.display = 'none'
  const timeout = setTimeout(() => {
    console.log('Timed out trying to clear Auth0 session.')
    document.body.removeChild(iframe)
  }, 10000)
  const eventHandler = () => {
    clearTimeout(timeout)
    removeEventListener('message', eventHandler)
    document.body.removeChild(iframe)
  }
  addEventListener('message', eventHandler)
  document.body.appendChild(iframe)
  iframe.setAttribute('src', `https://${domain}/v2/logout`)
}
```

I've confirmed that eventHandler does get called when the iframe loads, but I was also copying this somewhat blindly from runIframe in the first place, so if it looks wrong, LMK.

(Also, that permission error is weird, but if I look at my branch, your commits are on there, so perhaps it worked after all? 🤷🏼‍♂️)

stevehobbsdev commented:
@majelbstoat Looks good at first glance, let's get it in situ and have a play with it. 

> that permission error is weird, but if I look at my branch, your commits are on there

I just merged from master, perhaps those are the commits you're seeing. I also underestimated how helpful the `gh` CLI tool would be here it seems if you checkout a fork PR you can't just magically commit back to it 😂

balibebas commented:
Looking at the diff it's hard to see what changed due to code style pollution. Which begs the question, what's left?

stevehobbsdev commented:
@majelbstoat Will you be continuing your implementation here or should we close?

charsleysa commented:
> @majelbstoat I think we're going to need to change up the mechanism slightly. In testing `logoutSilently`, the 60 timeout will always be hit.
> 
> The way `runIframe` works is that it relies on Auth0 sending back a response that has some JavaScript in it that uses `postMessage` to return some data back to the SDK. We don't have this for the logout endpoint (it just returns a redirect back to the app), so `runIframe` never knows when it is supposed to complete.
> 
> I think we can use the logout endpoint as a kind of fire-and-forget call as there won't be any real useful response. Then we can clear the cache and log the user out locally. Unfortunately it probably means you can't use the `runIframe` method we have (or put a really short timeout on it and absorb the timeout error - bit nasty but would work).
> 
> Make sense?
> 
> I'd like to help push some changes here but I don't have permission to edit this PR it seems (when I try to push I get asked for my creds).

I wish to propose that two methods are logoutSilently are offered.
1. fire-and-forget style where there is no guarantee of success and is essentially based on waiting for some amount of time to pass and hope that it succeeded
2. custom hosted "web_message" page which is able to message back to our main window that the logout has completed successfully, with the url being passed to the logout endpoint via the `returnTo` parameter

If no `returnTo` parameter is specified, option 1 is automatically selected. If it is specified then option 2 is automatically selected. This would be the smoothest DX in my opinion.

For option 2, the page can return the following html:
<details>
<summary>open me to see example</summary>

```html
<!DOCTYPE html>
<html>

<head>
    <title>Logout Response</title>
</head>

<body>
    <script type="text/javascript">
        (function(window, document, undefined) {
            // Begin : these values rendered by server
            var redirectURI = "https://client.example.com"; // (NOT the returnTo value)
            var logoutResponse = {
                type: "logout_response",
                response: {
                    success: true
                }
            };
            // End
            var mainWin = (window.opener != window) ? window.opener : window.parent;
            mainWin.postMessage(authorizationResponse, redirectURI);
        })(this, this.document);
    </script>
</body>

</html>
```

</details>

Thoughts?

Also, I'm willing to takeover the development for this if @majelbstoat is no longer available.

adamjmcgrath commented:
Hi @majelbstoat - thanks for submitting this PR

After some discussions with the team, we've decided not to include a `logoutSilently` feature in the SDK, this is mainly because:

- we don't want to offer a logout feature which can't tell you if the logout succeeded or not (the fire and forget option)
- we  don't want to offer a logout feature which would require you to host a custom "web_message" response page, although we do see the merit in this approach and will either flesh this  out in a sample or blog post.

Apologies for the back and forth, and thanks again for you work on this PR
